oeps/oep-38-Data-Modeling.rst

```diff
@@ -0,0 +1,401 @@
+The application of these standards is the most basic level of support that all application in the OpenedX ecosystem
+should adhere. When creating new applications or models please ensure the models being created comform to the following.
+
+Use a numeric primary key
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  It is recommended to use use :code:`BigAutoField`_.
+.. _BigAutoField: https://docs.djangoproject.com/en/2.2/ref/models/fields/#bigautofield
+
+-  Do not use composite based primary keys. Use a primary key column.
+
+Have updated and created timestamps
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  The preferred method for doing this in OpenedX Django applications is to inherit the :code:`TimeStampedModel`_. class.
+.. _TimeStampedModel: https://django-model-utils.readthedocs.io/en/latest/models.html#timestampedmodel
+
+- Time should be stored in UTC time by setting :code:`USE_TZ=True`_. in your python config.
+.. _USE_TZ=True: https://docs.djangoproject.com/en/2.2/topics/i18n/timezones/#overview
+
+-  If for some reason you can not inherit from :code:`TimeStampedModel`_. use the following naming conventions:
+
+   -  Created date should be named: “created”
+
+   -  Updated date should be named: “modified”
+
+Foreign keys
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  Data should be joined using primary keys wherever possible
+
+-  Foreign keys should use a naming convention of :code:`<object_name>_id` where object name is the name of the table of the foreign key relationship.
+
+-  Do not join on things such as username, email address, or other dimensions of data that may change over time
+
+-  Do not join on PII
+
+- Joining between IDAs should be done by using an Universally unique identifier (UUID)
+
+- In Django use `Attributes for fields with relations`_. to identify and link models with relationships.
+.. __Attributes for fields with relations: https://docs.djangoproject.com/en/2.2/ref/models/fields/#module-django.db.models.fields.related
+
+History for models involved with enrollments, courses and course metadata, or data involving financial payments and transactions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  History using django-simple-history_. .. _django-simple-history: https://django-simple-history.readthedocs.io/en/latest/
+
+-  Remember to `backfill history`_. for existing models.
+
+.. _backfill history: https://django-simple-history.readthedocs.io/en/latest/quick_start.html#existing-projects
+
+-  Where Django simple history is not an option, the following data should be captured:
+
+   -  Fields that were changed
+
+   -  Date & time of the change
+
+   -  The foreign key of of the user who initiated the change
+
+Use the correct data type for a column
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  Don’t use a :code:`IntegerField` when a :code:`BooleanField` would do.
+
+-  Use :code:`BigIntegerField`_. for your foreign keys
```

Comment:
Is the period after `BigIntegerField`_ part of the syntax or a typo?

Reply:
It's part of the syntax to make that code block also a clickable link.